### ISSUE TITLE
Add flag to rclone check to output a flat list of 'diff' files

### ISSUE THREAD

**Issue by 88lex** [rclone/rclone#3264]
#### What is your current rclone version (output from `rclone version`)?
rclone v1.47.0-126-g44f64917-beta

#### What problem are you are trying to solve?
No problem. But would be great to be abe to run rclone check and create a flat list of files that exist on `only source` or `only destination`. That list could then be used for processing the differential files (move, delete, etc.).

Reference our discussion in this thread:
https://forum.rclone.org/t/rclone-check-option-to-output-list-of-differences-between-source-and-destination/10430

#### How do you think rclone should be changed to solve that?
Add a flag or flags that alter the standard output of rclone check to instead output flat files. 
Some or all of the following would be outstanding.
--source-only
--destination-only
--diff-list

Since the name of a file can be repeated, with different contents, within a file tree the output would ideally include the full location from the root of the remote:
/topfolder/subfolder/file1.txt
as opposed to only showing
file1.txt

**Comment by ncw** [comment]
So having flags which would output to files would be easy enough I think

    --source-only <filename>
    --destination-only <filename>
    --different <filename>
    --same <filename>

Where every file on the source or dest would end up in one of those files (if the flag was supplied).  Those files could then be fed back to rclone via the --files-from flag.

Alternatively (or as well) I could do something like this output to stdout which would be easy for scripting

```
+ extra_file_in_dst
- file_missing_in_dst
* file_differs_in_dst
= file_same_in_dist
```

**Comment by 88lex** [comment]
The above would be absolutely perfect for my use cases, thank you!

I'd be happy to help but my programming-fu is limited to very simple bash scripts. My apologies.

**Comment by mirfatif** [comment]
> ```
> + extra_file_in_dst
> - file_missing_in_dst
> * file_differs_in_dst
> = file_same_in_dist
> ```

And this may hopefully reduce my work to get a list of changed files and calculate their upload size before syncing:
```
rclone --dry-run /path/to/DIR GDrive:/DIR 2>&1 | 
  sed -n 's|.*NOTICE: |DIR/|; s|: Not copying .*||p' |
  xargs -I {} stat -c '%s %n' '{}' |
  awk '{ sum += $1; print } END { print sum }'
```
I'm not sure if there already exists some better solution to achieve the same.

**Comment by ncw** [comment]
I've made a prototype of this here

https://beta.rclone.org/branch/v1.52.2-131-g956c0d38-fix-3264-check-output-beta/ (uploaded in 15-30 mins)

Check has got these new flags

      --combined string   Make a combined report of changes to this file
      --differ string     Report all non-matching files to this file
      --dst-only string   Report all files missing from the source to this file
      --match string      Report all matching files to this file
      --src-only string   Report all files missing from the destination to this file

You can specify any of the file names as `-` which means output to stdout.

The `--differ`, `--dst-only`, `--match` and `--src-only` flags write file paths, one per line of type indicated.

So `--differ` will contain all files which are different on the source and dest, etc

The `--combined` file contains all file paths with a symbol at the front of the line to tell you what happened to it.

`=` means the files matched, `-` means the file was missing on the destination, `+` means the file was extra on the destination and `*` means that the file was present in source and destination but different.

Testing appreciated!

**Comment by ncw** [comment]
I had feed back that the flag names were too confusing so I've changed them here.

https://beta.rclone.org/branch/v1.52.2-136-g839d544a-fix-3264-check-output-beta/ (uploaded in 15-30 mins)

The `--differ`, `--missing-on-dst`, `--missing-on-src` and
`--src-only` flags write paths, one per line, to the file name (or
stdout if it is `-`) supplied. What they write is described in the
help below. For example `--differ` will write all paths which are
present on both the source and destination but different.

The `--combined` flag will write a file (or stdout) which contains all
file paths with a symbol and then a space and then the path to tell
you what happened to it. These are reminiscent of diff files.

- `= path` means path was found in source and destination and was identical
- `- path` means path was missing on the source, so only in the destination
- `+ path` means path was missing on the destination, so only in the source
- `* path` means path was present in source and destination but different.

Flags:

      --combined string         Make a combined report of changes to this file
      --differ string           Report all non-matching files to this file
      --match string            Report all matching files to this file
      --missing-on-dst string   Report all files missing from the destination to this file
      --missing-on-src string   Report all files missing from the source to this file

**Comment by mirfatif** [comment]
@ncw I want to get all files which differ or are missing on destination.
- Should I use `rclone --filter-from <filter-list> --diff --missing-on-dst <source> <remote>`?
- If I provide `--links` option and the file is a symlink does it return `.rclonelink` appended?
- I also want file size of all changed/new files and their total size. I pipe it to `xargs -I {} stat -c '%s %n' '{}' | awk '{ sum += $1; print } END { print sum }'`. Do you suggest a btter way?
- Which stable release is expected to have `--diff` and accompanied options?

**Comment by ncw** [comment]
> @ncw I want to get all files which differ or are missing on destination.
> 
> * Should I use `rclone --filter-from <filter-list> --diff --missing-on-dst <source> <remote>`?

You'd use `--differ file-name  --mising-on-dst filename` to get the list of files into the same file

> * If I provide `--links` option and the file is a symlink does it return `.rclonelink` appended?

Yes

> * I also want file size of all changed/new files and their total size. I pipe it to `xargs -I {} stat -c '%s %n' '{}' | awk '{ sum += $1; print } END { print sum }'`. Do you suggest a btter way?

It sounds like you are re-inventing `rclone sync` - maybe you could tell me more about what you want to acheive? `rclone sync --dry-run` will output those stats. Maybe `rclone check` should total up the number of bytes rclone would need to transfer to make the source and the dest identical?

> * Which stable release is expected to have `--diff` and accompanied options?

1.53

**Comment by mirfatif** [comment]
> You'd use `--differ file-name --mising-on-dst filename` to get the list of files into the same file

So if there is a long list of files, I need to mention every file name like: `--differ file1 --mising-on-dst file1 --differ file2 --mising-on-dst file2 ...`?

> It sounds like you are re-inventing `rclone sync` - maybe you could tell me more about what you want to acheive? `rclone sync --dry-run` will output those stats. Maybe `rclone check` should total up the number of bytes rclone would need to transfer to make the source and the dest identical?

I want this:
```
~$ tree .
.
├── 1
│   ├── file1
│   ├── file2
│   └── file3
└── 2

2 directories, 3 files
~$ rclone --dry-run sync 1 2 2>&1 | sed -u -n 's|.*NOTICE: |1/|; s|: Not copying .*||p' | xargs -I {} stat -c '%s %n' '{}' | awk '{ sum += $1; print } END { print sum }'
5242880 1/file1          <--- Individual size of file1
10485760 1/file2         <--- Individual size of file2
2097152 1/file3          <--- Individual size of file3
17825792                 <--- Total size of all files
```

**Comment by ncw** [comment]
> > You'd use `--differ file-name --mising-on-dst filename` to get the list of files into the same file
> 
> So if there is a long list of files, I need to mention every file name like: `--differ file1 --mising-on-dst file1 --differ file2 --mising-on-dst file2 ...`?

`rclone check` outputs the missing files into that `filename` you don't have to put them in as arguments.

> > It sounds like you are re-inventing `rclone sync` - maybe you could tell me more about what you want to acheive? `rclone sync --dry-run` will output those stats. Maybe `rclone check` should total up the number of bytes rclone would need to transfer to make the source and the dest identical?
> 
> I want this:

I meant what workflow are you trying to achieve?

**Comment by mirfatif** [comment]
Thanks for your patience.

> `rclone check` outputs the missing files into that `filename` you don't have to put them in as arguments.

Well that doesn't serve the purpose i.e. to get a plain file list which is easily machine processable. I have already been using `--dry-run` with `copy` or `sync` to achieve this. Using `check` improves nothing (if I'm not missing something that `check` can do).

> I meant what workflow are you trying to achieve?

If I understand you correctly, I run `rclone` bash scripts in a cron (or on Android *cron-like*) service on rootfs `/` with a `--filters-from` list. That's how I keep my Linux systems and Android devices backed up. There's no specfic requirement though (except that I want to check files with their sizes before syncing). I'm open to adaptation.

**Comment by ncw** [comment]
> > `rclone check` outputs the missing files into that `filename` you don't have to put them in as arguments.
> 
> Well that doesn't serve the purpose i.e. to get a plain file list which is easily machine processable. I have already been using `--dry-run` with `copy` or `sync` to achieve this. Using `check` improves nothing (if I'm not missing something that `check` can do).

What `rclone check --differs filename.txt` does is very similar to running `rclone sync --dry-run` and parsing the log. The output in filename.txt will be one filename per line so you don't need to parse logs.

However it isn't 100% compatible as `rclone check` will check checksums whereas `rclone sync` will check modtimes first by default.

> > I meant what workflow are you trying to achieve?
> 
> If I understand you correctly, I run `rclone` bash scripts in a cron (or on Android _cron-like_) service on rootfs `/` with a `--filter-from` list. That's how I keep my Linux systems and Android devices backed up. There's no specfic requirement though (except that I want to check files with their sizes before syncing). I'm open to adaptation.

So your desire is to do a sync, but before you run it to see exactly what files would be synced and how big they are? I think your current method using `rclone sync --dry-run` is probably the most accurate method for that.

**Comment by mirfatif** [comment]
> So your desire is to do a sync, but before you run it to see exactly what files would be synced and how big they are?

Exactly.

> I think your current method using `rclone sync --dry-run` is probably the most accurate method for that.

Thanks for your time.

**Comment by ncw** [comment]
This is all done now and merged into v1.53
